### PR TITLE
Use @actions/tool-cache

### PR DESCRIPTION
**Pull request by ericmj**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -21,6 +21,7 @@
   "dependencies": {
     "@actions/core": "1.10.0",
     "@actions/exec": "1.1.1",
+    "@actions/tool-cache": "^2.0.1",
```
Comment (paulo-ferraz-oliveira) [inline package.json:24]:
```suggestion
    "@actions/tool-cache": "2.0.1",
```

**Comment by ericmj** [comment]
@paulo-ferraz-oliveira I think this is ready for review.

Gleam and Rebar3 are still not using the tool-cache but I don't use them and don't want to deal with more merge conflicts so I am only doing Elixir and OTP in this PR. If someone wants to use to contribute the tool-cache for Gleam and Rebar3 that would be great.

I also learnt tool-cache is only caching between usages in a single workflow, it doesn't cache between different CI runs. So if we actually want proper caching we need to use `@actions/cache`. But tool-cache is still an improvement because it adds retries and removes the shell files and much of the different handling between linux and windows.

**Comment by starbelly** [comment]
> Gleam and Rebar3 are still not using the tool-cache but I don't use them and don't want to deal with more merge conflicts so I am only doing Elixir and OTP in this PR. If someone wants to use to contribute the tool-cache for Gleam and Rebar3 that would be great.

This is great. I think the main rub was OTP anyway.

**Comment by starbelly** [comment]
We have a failing job, I'd be in favor of removing it or altering the job so it doesn't test with rebar. I believe that version of rebar isn't supported anymore, at the very least that version of  OTP supported isn't supported by rebar itself.  If we really want to test OTP 20 with rebar, we should probably bump the rebar version in there, thoughts?

**Comment by paulo-ferraz-oliveira** [comment]
> If someone wants to use to contribute the tool-cache for Gleam and Rebar3 that would be great.

I created an issue to track this: https://github.com/erlef/setup-beam/issues/204. If somebody wants to give it a go, it's there.

**Comment by paulo-ferraz-oliveira** [comment]
> We have a failing job

Which one? It's probably one of those I recently fixed, but wanna be in sync with this comment.

File: package-lock.json [hunk 2]
```diff
@@ -9,7 +9,8 @@
       "dependencies": {
         "@actions/core": "1.10.0",
         "@actions/exec": "1.1.1",
-        "@actions/http-client": "^2.1.0",
+        "@actions/http-client": "2.1.0",
+        "@actions/tool-cache": "^2.0.1",
```
Comment (paulo-ferraz-oliveira) [inline package-lock.json:13]:
```suggestion
        "@actions/tool-cache": "2.0.1",
```

File: package.json (hunk 1 above)
Comment (paulo-ferraz-oliveira) [reply package.json:25]: I think you need to run `npm run build-dist` again...

File: src/installer.js [hunk 3]
```diff
@@ -10,28 +13,66 @@ const path = require('path')
...
+          `https://builds.hex.pm/builds/otp/${fullVersion}.tar.gz`,
+        )
+        const extractPath = await tc.extractTar(tarPath, undefined, [
+          'zx',
+          '--strip-components=1',
+        ])
+        cachedPath = await tc.cacheDir(extractPath, 'otp', fullVersion)
+      }
+
+      // TODO: Can we cache install?
```
Comment (paulo-ferraz-oliveira) [inline src/installer.js:40]: If this is not to be solved in the context of this pull request, could you create a task to track it, and delete the comment? I think there's a similar one below. What's the measurable improvement in caching the installation too?

**Review by paulo-ferraz-oliveira — APPROVED** [review]
Approved with minor comments:
* needs a rebase
* there's a comment-related comment you might want to look at, otherwise we can do it in the future too

File: src/installer.js (hunk 3 above)
Comment (ericmj) [reply src/installer.js:40]: There's no improvement unless you do multiple installations in one workflow run. I don't think anyone is doing that.

Comment (paulo-ferraz-oliveira) [reply src/installer.js]: Perfect!

**Comment by paulo-ferraz-oliveira** [comment]
@ericmj, if/when you have time do you think you could summarise "I also learnt tool-cache is only caching between usages in a single workflow, it doesn't cache between different CI runs. So if we actually want proper caching we need to use @actions/cache. But tool-cache is still an improvement because it adds retries and removes the shell files and much of the different handling between linux and windows." in a new issue, for later tackling? Thanks much.

**Comment by ericmj** [comment]
I don't know yet if it's a good idea to do that, still looking into it. I can't find other actions using `@actions/cache` to cache installations.

**Comment by starbelly** [comment]
> I don't know yet if it's a good idea to do that, still looking into it. I can't find other actions using `@actions/cache` to cache installations.

Late to the party here, but I do think that kind of caching should be left up to the user.

**Comment by paulo-ferraz-oliveira** [comment]
Released in `v1.16.0` / `v1.16`. Tag `v1` moved up...

**Pull request merged**